### PR TITLE
fix generate_proposal_labels in cascade-rcnn series model, test=develop

### PR DESCRIPTION
**Pull request by jerrywgz**
### PR types

Function optimization

### PR changes

OPs

### Describe

Fix generate_proposal_labels in cascade-rcnn series model and add max_overlap as input. 
This op is only used at the stage of training so it will have no effect on inference.
This op is only used in static mode so core.ops,xxx is not used as well.

The mAP of related models:
Cascade Faster RCNN is 40.8 and previous version is 40.9.
HTC is 42.9/37.0 and previous is 42.7/36.8

The difference on document is marked as below:
image
image
image

### CODE REVIEW

File: paddle/fluid/operators/detection/bbox_util.h [hunk 1]
```diff
@@ -149,5 +149,19 @@ void ClipTiledBoxes(const platform::DeviceContext& ctx,
   }
 }
 
+template <typename T>
+void MaxOverlaps(const framework::Tensor& iou,
+                 framework::Tensor* max_overlaps) {
+  const T* proposal_to_gt_overlaps = iou.data<T>();
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/bbox_util.h]: 作为util的通用函数，命名也需要更通用。 proposal_to_gt_overlaps -> iou_data

File: paddle/fluid/operators/detection/bbox_util.h [hunk 2]
```diff
@@ -149,5 +149,19 @@ void ClipTiledBoxes(const platform::DeviceContext& ctx,
   }
 }
 
+template <typename T>
+void MaxOverlaps(const framework::Tensor& iou,
+                 framework::Tensor* max_overlaps) {
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/bbox_util.h]: max_overlaps -> max_iou?

File: paddle/fluid/operators/detection/bbox_util.h [hunk 3]
```diff
@@ -149,5 +149,19 @@ void ClipTiledBoxes(const platform::DeviceContext& ctx,
...
+  const T* proposal_to_gt_overlaps = iou.data<T>();
+  int row = iou.dims()[0];
+  int col = iou.dims()[1];
+  T* max_overlaps_dt = max_overlaps->data<T>();
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/bbox_util.h]: max_overlaps_dt -> max_iou_data

File: python/paddle/fluid/layers/detection.py [hunk 4]
```diff
@@ -2651,25 +2653,29 @@ def generate_proposal_labels(rpn_rois,
         use_random(bool): Use random sampling to choose foreground and background boxes.
         is_cls_agnostic(bool): bbox regression use class agnostic simply which only represent fg and bg boxes.
         is_cascade_rcnn(bool): it will filter some bbox crossing the image's boundary when setting True.
+        max_overlap(Variable): Maximum overlap between each Input box and ground-truth.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: Input box  -> each proposal box  .. ?

File: python/paddle/fluid/layers/detection.py [hunk 5]
```diff
@@ -2651,25 +2653,29 @@ def generate_proposal_labels(rpn_rois,
...
+        return_max_overlap(bool): Whether return the maximum overlap between each Output box and ground-truth.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: 同上，Output box更精确的描述

File: python/paddle/fluid/layers/detection.py [hunk 6]
```diff
@@ -2611,7 +2611,9 @@ def generate_proposal_labels(rpn_rois,
                              class_nums=None,
                              use_random=True,
                              is_cls_agnostic=False,
-                             is_cascade_rcnn=False):
+                             is_cascade_rcnn=False,
+                             max_overlap=None,
+                             return_max_overlap=False):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py:2606]: max_overlap是否始终return，而不加这个参数

File: python/paddle/fluid/tests/test_detection.py [hunk 7]
```diff
@@ -289,40 +289,40 @@ def test_anchor_generator(self):
...
+            rpn_rois = fluid.data(
+                name='rpn_rois', shape=[4, 4], dtype='float32', lod_level=1)
+            gt_classes = fluid.data(
+                name='gt_classes', shape=[6], dtype='int32', lod_level=1)
+            is_crowd = fluid.data(
+                name='is_crowd', shape=[6], dtype='int32', lod_level=1)
+            gt_boxes = fluid.data(
+                name='gt_boxes', shape=[6, 4], dtype='float32', lod_level=1)
+            im_info = fluid.data(
+                name='im_info', shape=[1, 3], dtype='float32', lod_level=1)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/test_detection.py]: im_info的lod_level是0？

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc [hunk 8]
```diff
@@ -486,25 +540,36 @@ class GenerateProposalLabelsKernel : public framework::OpKernel<T> {
       Tensor gt_boxes_slice =
           gt_boxes->Slice(gt_boxes_lod[i], gt_boxes_lod[i + 1]);
       Tensor im_info_slice = im_info->Slice(i, i + 1);
+      Tensor max_overlap_slice;
+      if (is_cascade_rcnn) {
+        auto* max_overlap = context.Input<Tensor>("MaxOverlap");
+        max_overlap_slice =
+            max_overlap->Slice(rpn_rois_lod[i], rpn_rois_lod[i + 1]);
+      } else {
+        max_overlap_slice.mutable_data<T>(im_info_slice.dims(),
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/generate_proposal_labels_op.cc]: 这里的shape是和im_info_slice相同吗？ im_info_slice的shape是[1, 3]?

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc [hunk 9]
```diff
@@ -550,6 +617,12 @@ class GenerateProposalLabelsOpMaker : public framework::OpProtoAndCheckerMaker {
              "(Tensor), This input is a 2D Tensor with shape [B, 3]. "
              "B is the number of input images, "
              "each element consists of im_height, im_width, im_scale.");
+    AddInput("MaxOverlap",
+             "(LoDTensor), This input is a 1D LoDTensor with shape [N]."
+             "N is the number of Input(RpnRois), "
+             "each element is the maxoverlap between "
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/generate_proposal_labels_op.cc]: maxoverlap -> max overlap

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc [hunk 10]
```diff
@@ -33,6 +33,25 @@ void AppendRois(LoDTensor* out, int64_t offset, Tensor* to_add) {
...
+  const T* rpn_rois_dt = rpn_rois.data<T>();
+  const T* max_overlap_dt = max_overlap.data<T>();
+  int rois_num = max_overlap.numel();
+  keep->Resize({rois_num});
+  int* keep_data = keep->mutable_data<int>(ctx.GetPlace());
+  int keep_len = 0;
+  for (int i = 0; i < rois_num; ++i) {
+    if ((rpn_rois_dt[i * 4 + 2] - rpn_rois_dt[i * 4 + 0] + 1) > 0 &&
+        (rpn_rois_dt[i * 4 + 3] - rpn_rois_dt[i * 4 + 1] + 1) > 0 &&
+        max_overlap_dt[i] < 1.) {
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/generate_proposal_labels_op.cc:51]: add comments for why filter max_overlap_dt < 1.0

File: paddle/fluid/operators/detection/bbox_util.h (hunk 2 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/bbox_util.h]: Done

File: paddle/fluid/operators/detection/bbox_util.h (hunk 3 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/bbox_util.h]: Done

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc (hunk 10 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/generate_proposal_labels_op.cc:51]: Done

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc (hunk 8 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/generate_proposal_labels_op.cc]: Done, thx

File: paddle/fluid/operators/detection/generate_proposal_labels_op.cc (hunk 9 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/generate_proposal_labels_op.cc]: Done

File: python/paddle/fluid/layers/detection.py (hunk 6 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/detection.py:2606]: 这里希望可以兼容PaddleDetection先前的版本，如果始终return，https://github.com/PaddlePaddle/PaddleDetection/blob/release/0.4/ppdet/modeling/architectures/mask_rcnn.py#L125 计算 loss的位置会受到影响

File: python/paddle/fluid/layers/detection.py (hunk 4 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/detection.py]: Done

File: python/paddle/fluid/layers/detection.py (hunk 5 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/detection.py]: Done

File: python/paddle/fluid/tests/test_detection.py (hunk 7 above)
Comment (jerrywgz) [reply python/paddle/fluid/tests/test_detection.py]: Done, thx

**Review by qingqing01 — APPROVED** [review]

**Review by Superjomn — APPROVED** [review]
考虑到只训练使用，这里不做兼容要求

**Review by Heeenrrry — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
generate_proposal_labels is used in static graph

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**